src/terms/ewald.jl

```diff
@@ -116,23 +136,106 @@ function energy_forces_ewald(lattice::AbstractArray{T}, charges, positions;
+end
+
+# TODO: See if there is a way to express this with AD.
+function dynmat_ewald_recip(model::Model{T}, τ, σ; η=default_η(model.lattice),
+                            q=zero(Vec3{T})) where {T}
+    # Numerical cutoffs to obtain meaningful contributions. These are very conservative.
+    # The largest argument to the exp(-x) function
+    max_exp_arg = -log(eps(T)) + 5  # add some wiggle room
+
+    lattice       = model.lattice
+    recip_lattice = model.recip_lattice
+    # Precomputing summation bounds from cutoffs.
+    # In the reciprocal-space term we have exp(-||B G||^2 / 4η^2),
+    # where B is the reciprocal-space lattice, and
+    # thus use the bound  ||B G|| / 2η ≤ sqrt(max_exp_arg)
+    Glims = estimate_integer_lattice_bounds(recip_lattice, sqrt(max_exp_arg) * 2η)
+
+    charges   = T.(charge_ionic.(model.atoms))
+    positions = model.positions
+    @assert length(charges) == length(positions)
+    pτ = positions[τ]
+    pσ = positions[σ]
+
+    dynmat_recip = zeros(complex(T), length(q), length(q))
+    for G1 in -Glims[1]:Glims[1], G2 in -Glims[2]:Glims[2], G3 in -Glims[3]:Glims[3]
+        G = Vec3(G1, G2, G3)
+        if !iszero(G + q)
+            Gsqq = sum(abs2, recip_lattice * (G + q))
+            term = exp(-Gsqq / 4η^2) / Gsqq * charges[σ] * charges[τ]
+            term *= cis2pi(dot(G + q, pσ - pτ))
+            term *= (2T(π) * (G + q)) * transpose(2T(π) * (G + q))
+            dynmat_recip += term
+        end
+
+        (iszero(G) || σ ≢ τ) && continue
+        Gsq = sum(abs2, recip_lattice * G)
+
+        strucfac = sum(Z * cos2pi(dot(pσ - r, G)) for (r, Z) in zip(positions, charges))
+        dsum = charges[σ] * strucfac
+        dsum *= (2T(π) * G) * transpose(2T(π) * G)
+        dynmat_recip -= exp(-Gsq / 4η^2) / Gsq * dsum
+    end
+
+    # Amend `dynmat_recip` by proper scaling factors:
+    dynmat_recip *= 4T(π) / compute_unit_cell_volume(lattice)
+end
+
+# Computes the Fourier transform of the force constant matrix of the Ewald term.
+function dynmat_ewald(model::Model{T}; η=default_η(model.lattice), q=zero(Vec3{T})) where {T}
+    n_atoms = length(model.positions)
+    n_dim = model.n_dim
+
+    dynmat = zeros(complex(T), n_dim, n_atoms, n_dim, n_atoms)
+    # Real part
+    for τ in 1:n_atoms
+        for γ in 1:n_dim
+            displacement = zero.(model.positions)
+            displacement[τ] = StaticArrays.setindex(displacement[τ], one(T), γ)
+            real_part = -ForwardDiff.derivative(zero(T)) do ε
+                forces = energy_forces_ewald(model; η, ph_disp=ε .* displacement, q).forces
+                hcat(Array.(forces)...)
+            end
+
+            dynmat[:, :, γ, τ] = real_part[1:n_dim, :]
+        end
+    end
+    # Reciprocal part
+    for τ in 1:n_atoms
+        for σ in 1:n_atoms
+            dynmat[:, σ, :, τ] += dynmat_ewald_recip(model, σ, τ; η, q)
+        end
+    end
+    dynmat
+end
+
+function compute_dynmat(::TermEwald, scfres::NamedTuple; kwargs...)
```

Comment:
Not sure I like the fact, that the `scfres` appears here. For the other computational tasks (Hamiltonian, force) we "unfold" the scfres before passing it down to the terms. I'd suggest we do the same here.